*: Improve constraints retrieval in table descriptor

Xiang-Gu opened the pull request:
This PR tries to improve how we retrieve constraints in a table descriptor. Previously,
it was mostly legacy code carries over from a while ago and nothing hasn't really
changed.
The main change is to introduce `catalog.Constraint` interface, similar to `catalog.Index`
and `catalog.Column`, as the preferred interface for constraint in this layer. Previously,
we would directly expose the underlying protobuf descriptor.

Commit 1 (easy): Rename `catalog.ConstraintToUpdate` to `catalog.Constraint`. 
It's good that we already have an interface that is suitable to be used for our effort.

Commit 2 (easy): Added methods in just renamed `catalog.Constraint` interface for
index-backed-constraints (i.e. PRIMARY KEY or UNIQUE);

Commit 3 (easy): Let `tabledesc.index` struct implement `catalog.Constraint` interface
as we will use it for index-backed-constraints.

Commit 4 (easy): Add a method in `catalog.Constraint` that gets validity of the constraint.

Commit 5 (moderate): Add logic (`ConstraintCache`) to pre-compute all constraints in a
table, categorized by type and validity, so that we can readily retrieve them later. This is the same
idea/technique used for managing index and columns (in `IndexCache` and `ColumnCache`).

Commit 6 (easy): Introduce the new, preferred methods in `catalog.TableDescriptor` to
retrieve constraints from a table.

Commit 7 (easy): Refactor signature of the existing `FindConstraintWithID` method to use
the newly added interface and retrieval methods.

Informs: #90840 
(this PR can unblock #90840)

Release note: None

Xiang-Gu committed:
catalog, tabledesc: Rename an interface and its backing struct

This PR renames the following interface and struct:
  1. `catalog.ConstraintToUpdate` interface is now `catalog.Constraint`
  2. `tabledesc.constraintToUpdate` struct is now `tabledesc.constraint`

Xiang-Gu committed:
catalog: Refactor FindConstraintWithID method in TableDescriptor

This commit refactor `FindConstraintWithID` method inside the
`catalog.TableDescriptor` in the following two ways:
 1. Change the return type to the newly introduced `catalog.Constraint`
    instead of a protobuf struct;
 2. Change the implementation slightly to also search dropping
    constraints. Previously, it will skip those constraints.

The remaining changes in this commit are consequential changes due to
this function signature change. All of them are peripheral and does not
affect functionality.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/91704)
<!-- Reviewable:end -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 6 of 6 files at r1, 4 of 4 files at r2, 2 of 2 files at r3, 4 of 4 files at r4, 4 of 4 files at r5, 9 of 9 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91704#-:-NGr2QoS8jZz8ycqqekl:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @Xiang-Gu)

___
*[`pkg/sql/catalog/tabledesc/mutation.go` line 200 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91704#-NGr2-T63n_OePtioYOV:-NGr2-T7CP3bAnyGRRJc:b6e08lk) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb0aec52cddc31139eaa5fea21269f7af8083c64/pkg/sql/catalog/tabledesc/mutation.go#L200)):*
> ```Go
> 		return c.UniqueWithoutIndex().Validity
> 	} else {
> 		panic("unknown constraint type")
> ```

please panic an assertion failed error, the string will get redacted


<!-- Sent from Reviewable.io -->

postamar commented on pkg/sql/catalog/table_elements.go:
```diff
@@ -432,8 +438,19 @@ type ConstraintToUpdate interface {
 	UniqueWithoutIndex() descpb.UniqueWithoutIndexConstraint
 
+	// PrimaryKey returns the underlying index descriptor backing the UNIQUE constraint,
+	// if there is one.
```
nit: can you add a `TODO(postamar): return interface instead of protobuf` comment here and for `Unique()` please?

Xiang-Gu replied:
I will just do it here myself.

postamar commented on pkg/sql/catalog/tabledesc/constraint.go:
```diff
@@ -0,0 +1,238 @@
+	allActiveAndInactiveFKs []catalog.Constraint
+	allActiveFKs            []catalog.Constraint
+
+	allUniqueWithoutIndexs                  []catalog.Constraint
```
nit: `s/allUniqueWithoutIndexs/allUniqueWithoutIndexes`; same for 2 lines below.

Xiang-Gu replied:
done

postamar commented on pkg/sql/catalog/tabledesc/constraint.go:
```diff
@@ -0,0 +1,238 @@
+		dest = append(dest, c)
+		constraintIDsInDest[c.GetConstraintID()] = true
+		return dest
+	}
```
Is it at all possible for a constraint to present itself more than once?

Xiang-Gu replied:
yes, I knew of one scenario: when dropping a check constraint, we would find that constraint in the `Checks` slice, mark its validity to `DROPPING` *without* removing it from the `Checks` slice. We would also enqueue a mutation for this check constraint with `DROP` direction. Here, this dropping check constraint will be found both in `Checks` slice and on `Mutations` slice.

postamar replied:
Thanks! I feared it would be something like this.

postamar reviewed (commented):
Ok, so, this is sort of OK, but it requires a fair bit of follow-up work, which I'll be happy to do.

Xiang-Gu commented on pkg/sql/catalog/tabledesc/constraint.go:
```diff
@@ -0,0 +1,238 @@
+	}
+
+	return &c
+}
```
sgtm, I will leave it as is for this PR.

Xiang-Gu commented on pkg/sql/catalog/descpb/structured.proto:
```diff
@@ -534,6 +541,16 @@ message ConstraintToUpdate {
+  // - VALIDATING when the index is being added (i.e. live on mutation slice with ADD direction);
+  // - DROPPING when the index is being dropped (i.e. live on mutation slice with DROP direction);
+  // - UNVALIDATED in no case!
+  optional ConstraintValidity index_backed_constraint_validity = 9 [(gogoproto.nullable) = false];
```
I will change as your suggested! However, I think we cannot get rid of `ConstraintToUpdateDesc()` because this is needed if we want to modify the underlying the protobuf. Now with `index` implementing `Constraint`, we will need to add another method, maybe called `IndexDesc() *descpb.IndexDescriptor`, if the to-be-modified constraint is backed by an index.

One such scanrio of needing to modify the underlying protobuf is setting the name of a constraint. This will show up in both RENAME CONSTRAINT or, more generally, the adding/dropping path of a ConstraintName element.

Xiang-Gu commented on pkg/sql/catalog/table_elements.go:
```diff
@@ -432,8 +438,19 @@ type ConstraintToUpdate interface {
 	// there is one.
 	UniqueWithoutIndex() descpb.UniqueWithoutIndexConstraint
 
+	// PrimaryKey returns the underlying index descriptor backing the UNIQUE constraint,
```
done

Xiang-Gu committed:
*: Implement constraints retrieval from a table with caching

1. We let `tabledesc.index` struct implement `catalog.Constraint` interface
so that it is used for any index-backed-constraints (PRIMARY KEY or
UNIQUE);

2. We implement a `constraintCache` inside a table wrapper that was
pre-populated and is ready for re-use. It allows us to easily retrieve
various sets of constraints (by kind and/or by validity) from the table.

3. We expose three methods in `catalog.TableDescriptor` interface to
retrieve various constraints in this table using this cache.

postamar reviewed (approved):
LGTM, nicely done.

Xiang-Gu commented:
TFTR!

bors r+

Xiang-Gu commented:
bors r-

Some pieces are missing and let me fill them up.

Xiang-Gu commented:
Now I think it's ready to be merged again!

bors r+
